### DISCUSSION TITLE
In entity class, is there a way to convert values prior to inserting them into DB?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2482] [Q&A] [answered]
I need this for backward compatibility and making my coding easier.

Something like convert Booleans to `{0,1}` and insert them as `number` into the DB. I’d work with Booleans within the entity class, but it would insert integers into the DB.

Something like this (this is not working, obviously):

```ts
@Property(val => +!!val, {columnType: 'SMALLINT'})
val: Boolean
```

**Comment by B4nan** [comment, marked as answer]
https://mikro-orm.io/docs/custom-types

**Reply by tukusejssirs** [reply]
Thanks!

Just for future reference:

```ts
import {Type} from '@mikro-orm/core'

export class BooleanNumber extends Type<boolean, number> {
	convertToDatabaseValue(value: boolean): number {
		return +!!value
	}

	convertToJSValue(value: 0 | 1): boolean {
		return Boolean(value)
	}
}
```

**Reply by tukusejssirs** [reply]
@B4nan, I tried to add default values to the entity, but `npx mikro-orm schema:create -r --schema m1`, it complains that I tried to add default Boolean values while setting the column type to `SMALLINT`.

I tried to use `default: 0`, `default: false`, `propertyName = false` and `propertyName = 0`, but it alway fails with this error.

Note that I use the `BooleanNumber` type from above.

**Update:** Could a DTO intefere? I have there set the default values too (as Booleans).

**Update 2:** Is there any other cache besides `temp/` folder? I tried to delete it prior to generating the schema, but it didn’t solve anything.

**Update 3:**

Something else must going on, because even after I removed default values definition from both the entity and the DTO, it throws the same error.

Here’s the error:

```ts
DriverException: create table "m3"."status_interval" ("col" SMALLINT not null default false); - column "col" is of type smallint but default expression is of type boolean
    at PostgreSqlExceptionConverter.convertException ({repo_root}/node_modules/@mikro-orm/core/dist/platforms/ExceptionConverter.js:8:16)
    at PostgreSqlExceptionConverter.convertException ({repo_root}/node_modules/@mikro-orm/postgresql/dist/PostgreSqlExceptionConverter.js:40:22)
    at PostgreSqlDriver.convertException ({repo_root}/node_modules/@mikro-orm/core/dist/drivers/DatabaseDriver.js:177:54)
    at {repo_root}/node_modules/@mikro-orm/core/dist/drivers/DatabaseDriver.js:181:24
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async SchemaGenerator.execute ({repo_root}/node_modules/@mikro-orm/knex/dist/schema/SchemaGenerator.js:285:13)
    at async SchemaGenerator.createSchema ({repo_root}/node_modules/@mikro-orm/knex/dist/schema/SchemaGenerator.js:28:9)
    at async Function.handleSchemaCommand ({repo_root}/node_modules/@mikro-orm/cli/dist/commands/SchemaCommandFactory.js:86:13)

previous error: create table "m3"."status_interval" ("col" SMALLINT not null default false); - column "col" is of type smallint but default expression is of type boolean
    at Parser.parseErrorMessage ({repo_root}/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket ({repo_root}/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse ({repo_root}/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> ({repo_root}/node_modules/pg-protocol/src/index.ts:7:48)
    at Socket.emit (node:events:390:28)
    at Socket.emit (node:domain:475:12)
    at addChunk (node:internal/streams/readable:324:12)
    at readableAddChunk (node:internal/streams/readable:297:9)
    at Socket.Readable.push (node:internal/streams/readable:234:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:199:23) {
  length: 182,
  severity: 'ERROR',
  code: '42804',
  detail: undefined,
  hint: 'You will need to rewrite or cast the expression.',
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'heap.c',
  line: '3065',
  routine: 'cookDefault'
}
```

**Update 4:** Disregard all of this. There was another entity which uses `BooleanNumber`. :man_facepalming: Anyway, I’ve learnt that the `default` config of the `@Property()` decorator must have a valid value based on the `columnType` type, and therefore it should not be based on the _property_ data type. For example:

```ts
@Property({type: BooleanNumber, columnType: 'SMALLINT', default: 0})
col: boolean
```

**Comment by tukusejssirs** [comment]
@B4nan, I created another custom type (`timestamptz(3)` in the DB and `number` with timestamp in ms on the JS side; this is required because of memory utilisation optimatisation).

I have encountered an issue that when I insert some data into the DB, it returns the data with `Date` object instead of `number` (timestamp in ms). Is this a bug or is it intended behaviour?

```ts
// Entity
@Entity({schema: '*', tableName: 'table'})
@Index({properties: ['id'], type: 'btree'})
@Index({properties: ['startTime'], type: 'btree'})
export class Ent {
  @PrimaryKey()
  id!: number

  @IsInt()
  @Property({type: DateNumber, columnType: 'TIMESTAMPTZ(3)', defaultRaw: 'NOW()', fieldName: 'start_time'})
  startTime = new Date().valueOf()

  constructor(dto: Dto) {
    this.startTime = createStatusIntervalDto.startTime
  }
}

// DTO
export class CreateStatusIntervalDto {
  @IsInt()
  @IsOptional()
  id?: number

  @IsInt()
  startTime = new Date().valueOf()
}

// Custom type
export class DateNumber extends Type<number, Date> {
  convertToDatabaseValue(value: number): Date {
    return new Date(value)
  }

  convertToJSValue(value: Date): number {
    return value.getTime()
  }
}

// Inserting some data into the DB
const qb = this.orm.em.createQueryBuilder(Ent)
qb.insert(dto)
const data = await qb.execute('get')

// And `data.startTime` is a Date object instead of number
```

**Reply by B4nan** [reply]
https://github.com/brianc/node-pg-types

**Reply by tukusejssirs** [reply]
Thanks, @B4nan. I tried the following code, but it fails, as the parameter of the callback of `setTypeParser()` needs to be a string.

```ts
import {Type} from '@mikro-orm/core'
import {types} from 'pg'

export class DateNumber extends Type<number, Date> {
  convertToDatabaseValue(value: number): Date {
    return new Date(value)
  }

  convertToJSValue(value: Date): number {
    const timestamptzOid = 1184
    types.setTypeParser(timestamptzOid, (value) => value.getTime())

    // Of course, the following won't work
    return value.getTime()
  }
}
```

I even tried to use `value.toISOString()` (or even '2022-01-28T16:04:38:874Z')as the cb param, but it fails with `Unaught SyntaxError: Invalid destructuring assignment target`. Should use a different OID?

**Reply by B4nan** [reply]
This is a global operation, nothing you should put into a custom type handler. Note that you will actually need a custom type that will handle date conversion if you want it somewhere else.

**Reply by tukusejssirs** [reply]
Could you please tell me where and how should I define that _custom type that will handle date conversion_?
